src/storage/src/wal.rs

```diff
@@ -0,0 +1,385 @@
+                self.id, self.current_sequence_number
+            )
+        })?;
+        self.current_file.flush().with_context(|| {
+            format!(
+                "failed to flush write to relation: {} wal segment {}",
+                self.id, self.current_sequence_number
+            )
+        })?;
+        self.current_bytes_written += len;
+        self.total_bytes_written += len;
+
+        Ok(())
+    }
+
+    /// Write a set of (Row, Timestamp, Diff) updates to the write-ahead log.
+    fn write(&mut self, updates: &[Update]) -> Result<(), anyhow::Error> {
+        // TODO: The allocation discipline for writes could probably be a lot
+        // better. Each WAL writer could keep a buffer that it writes into
+        // for everything and fall back to allocating stuff if we run out of
+        // space.
+        let mut buf = Vec::new();
+        for update in updates {
+            encode_update(&update.row, update.timestamp, update.diff, &mut buf)?;
+        }
+
+        self.write_inner(&buf)
+    }
+
+    fn write_progress_inner(&mut self, timestamp: Timestamp) -> Result<(), anyhow::Error> {
+        let mut buf = Vec::new();
+        encode_progress(timestamp, &mut buf)?;
+        self.write_inner(&buf)
+    }
+
+    /// Write a progress (alternatively, timestamp closure) message to the write-ahead
+    /// log.
+    ///
+    /// Optionally also switch to a new log segment file if the current log segment file
+    /// grows too large.
+    fn write_progress(&mut self, timestamp: Timestamp) -> Result<(), anyhow::Error> {
+        self.write_progress_inner(timestamp)?;
+
+        // We only want to rotate the log segments at progress messages
+        // so that we can extract batches.
+        if self.current_bytes_written > MAX_LOG_SEGMENT_SIZE {
+            self.rotate_log_segment()?;
+            // Write the progress update to the new file to make sure we
+            // know the `lower` for the new file
+            self.write_progress_inner(timestamp)?;
+        }
+
+        Ok(())
+    }
+
+    /// Switch to writing a new log segment file.
+    ///
+    /// Mark the previous one as finished by prepending `-final` to its name.
+    fn rotate_log_segment(&mut self) -> Result<(), anyhow::Error> {
+        let old_file_path = self
+            .base_path
+            .join(format!("log-{}", self.current_sequence_number));
+        let old_file_rename = self
+            .base_path
+            .join(format!("log-{}-final", self.current_sequence_number));
+        let new_file_path = self
+            .base_path
+            .join(format!("log-{}", self.current_sequence_number + 1));
+
+        // First lets open the new file
+        let new_file = create_log_segment(&new_file_path)?;
+        let old_file = std::mem::replace(&mut self.current_file, new_file);
+
+        // Let's close the file descriptor from the old file and mark it as final.
+        old_file.sync_all().with_context(|| {
```

Comment:
A sync is needed after every write or the file needs to be opened with the appropriate flag, and neither seems to be the case.

The files are opened like this:

```
[pid 3620036] openat(AT_FDCWD, "mzdata/table_wals/u1", O_RDONLY|O_NONBLOCK|O_CLOEXEC|O_DIRECTORY <unfinished ...>
```

and are written with ```write``` with no explicit syncing of any kind:

```
[pid 3620059] write(13, "\0\0\0\1\0\0\1w\355b\0\320", 12) = 12
[pid 3620059] write(13, "\0\0\0\1\0\0\1w\355b\4\270", 12) = 12
[pid 3620059] write(13, "\0\0\0\1\0\0\1w\355b\10\241", 12) = 12
[pid 3620059] write(13, "\0\0\0\1\0\0\1w\355b\f\211", 12) = 12
[pid 3620059] write(13, "\0\0\0\1\0\0\1w\355b\20r", 12) = 12
[pid 3620059] write(13, "\0\0\0\1\0\0\1w\355b\24Z", 12) = 12
[pid 3620059] write(13, "\0\0\0\1\0\0\1w\355b\30C", 12) = 12
[pid 3620059] write(13, "\0\0\0\1\0\0\1w\355b\34+", 12) = 12
[pid 3620059] write(13, "\0\0\0\1\0\0\1w\355b \24", 12) = 12
[pid 3620059] write(13, "\0\0\0\1\0\0\1w\355b#\374", 12) = 12
[pid 3620059] write(13, "\0\0\0\1\0\0\1w\355b'\344", 12) = 12
[pid 3620059] write(13, "\0\0\0\1\0\0\1w\355b+\315", 12) = 12
```
I guess that at this stage any syncing is OK. Group commits, doublewrite buffers, different sync approaches per OS/filesystem, etc. are all in the future.

Reply:
Interesting, I thought the call to `flush` that we do after every write would involve a sync

Reply:
So ok. the same file will be opened twice once by the WAL writer -- in that case I want the open syscall to have `O_CREAT | O_APPEND | O_SYNC | O_WRONLY` - i think that combination of flags gives the semantics we want?

and then once by the Compacter thread, which should open it with `O_RDONLY` I think the open call you've traced is one made by the compacter thread?

I don't have easy access to a linux box rn but I will document the intention here and double check on a linux box later

Reply:
Yes, O_SYNC should do the trick, but I do not see it . Here is the open by the WAL writer:

```
[pid 3789190] openat(AT_FDCWD, "mzdata/table_wals/u1/log-0", O_WRONLY|O_CREAT|O_EXCL|O_APPEND|O_CLOEXEC, 0666 <unfinished ...>
```

O_APPEND is used, so I assume it is the writer, but no O_SYNC.

Reply:
we chatted about this on slack and then I realized I was being silly

1. the rust api for opening files doesn't give me an easy way to open with `O_SYNC` 
2. but ... I can just call `File::sync_all` which should turn into a `fsync`(I think realistically the `sync_data` method probably calls `fdatasync` but for now this is fine)

so I've added a commit to do that.

Reply:
I'm going to add a todo here, remove the sync calls and call it a day for this, unless someone has a quick and brilliant solution